### DISCUSSION TITLE
Immutable vs Mutable

### DISCUSSION THREAD

**Discussion by Hellwoodoo** [Quillraven/Fleks#136] [Q&A]
So, was just wondering @Quillraven, what your thoughts are on immutable vs mutable components - and which you would recommend with Fleks. I intend to use this in a game engine with synchronous calls - but was wondering what your thoughts/advice are on this. Thanks!

**Comment by Quillraven** [comment]
I'd need more context for your use-case but usually components are mutable in my cases.

E.g. if I have an AnimationComponent then it contains a timer for the animation and this timer gets updated every frame to also then render the correct frame of the animation itself.

Some components can have immutable stuff like e.g. a GraphicComponent can have a reference to the Sprite class in LibGDX. Usually, there is no need for the sprite to change for an entity so that can be constructor argument of the GraphicComponent which does not change. Something like:

```kotlin
data class Graphic(val sprite: Sprite) : Component // ...

data class Animation(val animation: Animation2D<TextureRegion>) : Component<Animation> {
  var aniTimer: Float = 0f

  // ...
}
```

One more thing: Tags (=components without properties) are imo always immutable. I realize them either as enum values or as object classes.

**Reply by Hellwoodoo** [reply]
Interesting. Is it even possible for multiple threads to act together in a system?

**Reply by Hellwoodoo** [reply]
Additionally, how do you handle the case where you would like a texture component to be shared by multiple entities - to conserve resources?

**Reply by Quillraven** [reply]
If you sync it correctly then yes 😉
I am not a concurrency expert but I see two scenarios:

One scenario is like a runblocking call in your system's onUpdate call and you run that single system's code concurrently (=multiple coroutines inside runblocking).

The other scenario is that your systems itself run concurrently. I cannot remember but I think you cannot override the world's update method. So most likely you'd need your own ConcurrentSystem that then runs your systems concurrently internally in its update method.


But again, I am not an expert here. If you have feedback or suggestions to make your life easier in that regard feel free to open an issue or PR 😊

**Reply by Quillraven** [reply]
From what I know you don't do such things in ECS. One component is always related to one entity. Imo a component is not a big overhead. It is basically defining a Texture field on your entity object. The texture itself just exists once in your game but multiple entities can point to it.

**Reply by Hellwoodoo** [reply]
Yeah I guess it would be much easier to point to paths for textures, meshes, rather than moving around post-imported components with big arrays

**Reply by Quillraven** [reply]
That usually makes serialization also easier because a string of a path is easier to store and restore than a Sprite or 3D mesh.

However, it is imo still totally fine to keep a reference to such objects on the component itself and exclude them from the serialization process.

Whatever works best for you and is easier to use 🙂

**Reply by Hellwoodoo** [reply]
Thank @Quillraven for answering all these questions. Just another question, how exactly would you  implement an ECS with one-to-many relationships. Say for example you have a mesh which can contain a material, which can contain multiple textures. It feels incorrect to refer to components between each other - rather than in an overarching manner - but would this not necessitate the root mesh to be defined as an entity? 

Additionally - with Fleks - is it possible to create two components of the same type (e.g. local and world transform), or am I misunderstanding something? **Edit**: wait nvm this seems to be addressed by using advanced type

**Reply by Quillraven** [reply]
Unfortunately, I have no experience with 3D stuff. I did something 20 years ago with C++ and Irrlicht engine but that is too long ago 😅

Also, ECS is just one way to solve certain things in your game. If it does not feel natural for you to use then I'd stick with the normal approach that you use. However, for me personally it took a few games to develop until I understood better how to use components and systems. So you just might need to get some more experience with it to see the benefits 🙂

Having said that, let me give you an approach that I usually use which might help you.
First, an entity is just a unique ID. Nothing crazy about it and very simple. It like the identifier in your "components database" to know what kind of things it can do.

E.g. in my typical 2D games I have at least following components:
- Transform: to know where my entity is located and what size and boundaries it has. Usually just some floats
- Move: to give an entity the ability to move. The MoveSystem then Updates the Transform component values
- Graphic: to give an entity a graphical representation on the screen. Normally I use a Sprite here which also allows me to transform and tint the texture behind the Sprite. The position and size of the Sprite are coming from the Tramsform component which is done in a Graphic system that also renders the sprites


This is a very simple and basic setup. In your case it might make sense to have a Mesh component that has the information you mentioned like mesh, material, textures, etc.. If this is all needed to give an entity a "3D behavior" then imo it makes sense to put it into a single component.

If an entity can have multiple meshes them it might make sense to have multiple component types for your Mesh component. Like a "body" Mesh and a "head" mesh type or something like that. That way you can reuse your Mesh component data multiple times and link it to a single entity.

I think the wiki has an example for a Graphic component that has two types for a background and foreground.


Hope that helps you!

**Reply by Hellwoodoo** [reply]
Thanks, another question 😀, but how would you define an entity to be used in a world say as a camera, without needing to iterate through all entities to find the specific camera tag?

**Reply by Quillraven** [reply]
Not sure if I understand the question but systems are meant to be used to update entities with specific component configuration.

So if you want to do something with entities that have a CameraTag then just create an IteratingSystem for it. Fleks keeps track internally for such entities and will call the onTickEntity method of the system.

**Reply by Hellwoodoo** [reply]
More so want to use the properties of a specific entity (which may change) based on a tag - and use its component values in other systems. I am thinking of using an overarching system - but was wondering if there are better ways.
